### PR TITLE
[HttpClient] relax auth bearer format requirements

### LINKED ISSUE #38609: Unable to use `auth_bearer` option with Login with Amazon [HttpClient]
**Description**  
Unable to use `auth_bearer` option in `HttpClient` with [Login with Amazon](https://developer.amazon.com/apps-and-games/login-with-amazon) access token.

**How to reproduce**  
The problem is that a Login with Amazon access token contains the following character `|` which is not from the base64 alphabet. 

In order to reproduce you just need to pass a token like this:
```php
$response = $httpClient->request('GET', '/user/profile', [
   'auth_bearer' => 'Atza|IwEBIP_aN-yCg-udSr1lQvXHMzVwobs...', // Note that this is not the full token
]);
```
The error is pretty straightforward:
`Option "auth_bearer" must be a string containing only characters from the base 64 alphabet,  invalid string given.` 

**Possible Solution**  
Allow the `|` character in the regex defined in `HttpClientTrait` with something like `'{^[-._=~+/|0-9a-zA-Z]++$}'` or allow to pass an option to disable this check.

**Additional context**  
An example token looks like the following:
`Atza|IwEBIP_aN-yCg-udSr1lQvXHMzVwobsXt23OYs9ipZHq26ccXPZkR889oP5R7jKWs36f5Si63wlZX-7ijWek97vv8eygxGXExy3M7cGp959U98Z...`

I know that the easiest solution to this problem is to use the `Authorization` header and forget about it. But I secretly want to allow tokens with the `|` character. Why do we need this check anyways?

```php
$response = $httpClient->request('GET', '/user/profile', [
    'headers' => [
        'Authorization' => 'Bearer Atza|IwEBIP_aN-yCg-udSr1lQvXHMzVwobs...', // Note that this is not the full token
    ],
]);
```

If one of the provided solutions is okay I want to make a PR.

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #38609
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/HttpClient/HttpClientTrait.php [hunk 1]
```diff
@@ -110,7 +110,7 @@ private static function prepareRequest(?string $method, ?string $url, array $opt
             throw new InvalidArgumentException(sprintf('Option "auth_basic" must be string or an array, "%s" given.', \gettype($options['auth_basic'])));
         }
 
-        if (isset($options['auth_bearer']) && (!\is_string($options['auth_bearer']) || !preg_match('{^[-._=:~+/0-9a-zA-Z]++$}', $options['auth_bearer']))) {
+        if (isset($options['auth_bearer']) && (!\is_string($options['auth_bearer']))) {
             throw new InvalidArgumentException(sprintf('Option "auth_bearer" must be a string containing only characters from the base 64 alphabet, '.(\is_string($options['auth_bearer']) ? 'invalid string given.' : '"%s" given.'), \gettype($options['auth_bearer'])));
```
Comment (ro0NL) [inline src/Symfony/Component/HttpClient/HttpClientTrait.php]: redundant brackets + outdated message IIUC :)

**Comment by wouterj** [comment]
Thank you Christian.

File: src/Symfony/Component/HttpClient/HttpClientTrait.php (hunk 1 above)
Comment (xabbuh) [reply src/Symfony/Component/HttpClient/HttpClientTrait.php]: fixed

Comment (ro0NL) [reply src/Symfony/Component/HttpClient/HttpClientTrait.php]: still redundant brackets :P

**Comment by wouterj** [comment]
I just found https://github.com/symfony/symfony/pull/30561 (which introduced this regex) : 

> Better be sure CR/LF/etc cannot be passed inside raw header values, opening potential security risks.

Should we still check for these characters, to make sure they are not inside the bearer?

File: src/Symfony/Component/HttpClient/HttpClientTrait.php [hunk 2]
```diff
@@ -110,8 +110,8 @@ private static function prepareRequest(?string $method, ?string $url, array $opt
             throw new InvalidArgumentException(sprintf('Option "auth_basic" must be string or an array, "%s" given.', \gettype($options['auth_basic'])));
         }
 
-        if (isset($options['auth_bearer']) && (!\is_string($options['auth_bearer']) || !preg_match('{^[-._=:~+/0-9a-zA-Z]++$}', $options['auth_bearer']))) {
-            throw new InvalidArgumentException(sprintf('Option "auth_bearer" must be a string containing only characters from the base 64 alphabet, '.(\is_string($options['auth_bearer']) ? 'invalid string given.' : '"%s" given.'), \gettype($options['auth_bearer'])));
+        if (isset($options['auth_bearer']) && !\is_string($options['auth_bearer'])) {
+            throw new InvalidArgumentException(sprintf('Option "auth_bearer" must be a string , "%s" given.', \gettype($options['auth_bearer'])));
```
Comment (chalasr) [inline src/Symfony/Component/HttpClient/HttpClientTrait.php]: extra space before comma

File: src/Symfony/Component/HttpClient/HttpClientTrait.php [hunk 3]
```diff
@@ -110,8 +110,8 @@ private static function prepareRequest(?string $method, ?string $url, array $opt
             throw new InvalidArgumentException(sprintf('Option "auth_basic" must be string or an array, "%s" given.', \gettype($options['auth_basic'])));
         }
 
-        if (isset($options['auth_bearer']) && (!\is_string($options['auth_bearer']) || !preg_match('{^[-._=:~+/0-9a-zA-Z]++$}', $options['auth_bearer']))) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/HttpClientTrait.php:113]: let's keep `[\x21-\x7E]` as a check?

**Comment by nicolas-grekas** [comment]
Thank you @xabbuh.

**Pull request merged**